File: leetcode/169_MajorityElement.java
class Solution {
    public int majorityElement(int[] nums) {
        
        // Boyer–Moore majority vote algorithm

        int candidate = 0;
        int count = 0;

        for (int num : nums){

            if (count == 0) candidate = num;
            if (candidate == num) count ++;
            else count --;
        }

        return candidate;
    }
}

File: leetcode/169_MajorityElement.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:

        count = 0
        candidate = 0

        for i in range(len(nums)):

            if count == 0:
                candidate = nums[i]
                count += 1
            elif candidate == nums[i]:
                count += 1
            elif candidate != nums[i]:
                count -= 1

        return candidate     
            